src/Type/TypeAliasResolver.php

```diff
@@ -11,45 +11,124 @@ class TypeAliasResolver
+	 * @param array<string, string> $globalTypeAliases
 	 */
 	public function __construct(
-		array $aliases,
+		array $globalTypeAliases,
 		TypeStringResolver $typeStringResolver,
 		ReflectionProvider $reflectionProvider
 	)
 	{
-		$this->aliases = $aliases;
+		$this->globalTypeAliases = $globalTypeAliases;
 		$this->typeStringResolver = $typeStringResolver;
 		$this->reflectionProvider = $reflectionProvider;
 	}
 
-	public function hasTypeAlias(string $aliasName): bool
+	public function hasTypeAlias(string $aliasName, ?string $classNameScope): bool
 	{
-		return array_key_exists($aliasName, $this->aliases);
+		$hasGlobalTypeAlias = array_key_exists($aliasName, $this->globalTypeAliases);
+		if ($hasGlobalTypeAlias) {
+			return true;
+		}
+
+		if ($classNameScope === null || !$this->reflectionProvider->hasClass($classNameScope)) {
+			return false;
+		}
+
+		$classReflection = $this->reflectionProvider->getClass($classNameScope);
+		$localTypeAliases = $classReflection->getTypeAliases();
+		return array_key_exists($aliasName, $localTypeAliases);
 	}
 
 	public function resolveTypeAlias(string $aliasName, NameScope $nameScope): ?Type
 	{
-		if (!array_key_exists($aliasName, $this->aliases)) {
+		return $this->resolveLocalTypeAlias($aliasName, $nameScope)
+			?? $this->resolveGlobalTypeAlias($aliasName, $nameScope);
+	}
+
+	private function resolveLocalTypeAlias(string $aliasName, NameScope $nameScope): ?Type
+	{
+		$className = $nameScope->getClassName();
+		if ($className === null) {
+			return null;
+		}
+
+		// prevent infinite recursion
+		if (array_key_exists($className, $this->resolvingClassTypeAliases)) {
+			return null;
+		}
+
+		$this->resolvingClassTypeAliases[$className] = true;
+
+		if (!$this->reflectionProvider->hasClass($className)) {
+			return null;
+		}
+
+		$classReflection = $this->reflectionProvider->getClass($className);
+		$localTypeAliases = $classReflection->getTypeAliases();
+
+		unset($this->resolvingClassTypeAliases[$className]);
+
+		if (!array_key_exists($aliasName, $localTypeAliases)) {
+			return null;
+		}
+
+		$aliasNameInClassScope = $className . '::' . $aliasName;
+
+		if (array_key_exists($aliasNameInClassScope, $this->resolvedLocalTypeAliases)) {
+			return $this->resolvedLocalTypeAliases[$aliasNameInClassScope];
+		}
+
+		if ($this->reflectionProvider->hasClass($nameScope->resolveStringName($aliasName))) {
+			throw new \PHPStan\ShouldNotHappenException(sprintf('Type alias %s already exists as a class in scope of %s.', $aliasName, $className));
+		}
+
+		if (array_key_exists($aliasName, $this->globalTypeAliases)) {
+			throw new \PHPStan\ShouldNotHappenException(sprintf('Type alias %s used in scope of %s already exists as a global type alias.', $aliasName, $className));
+		}
+
+		if (array_key_exists($aliasNameInClassScope, $this->inProcess)) {
+			throw new \PHPStan\ShouldNotHappenException(sprintf('Circular definition for type alias %s in scope of %s.', $aliasName, $className));
```

Comment:
Throwing these exceptions causes internal errors. Instead it should be skipped here (`return null`) and reported in a new rule that checks these facts.

Reply:
Well, I've just preserved the behaviour that was already [there](https://github.com/phpstan/phpstan-src/blob/564b966a31dc339568aaa97c3bee6f195608e7cd/src/PhpDoc/TypeAlias/TypeAliasesTypeNodeResolverExtension.php#L57-L63) :) but ok, I'll try and think of a rule to check for these errors

Reply:
The difference is that before it was a user error in config but now it's an error in analysed source code (possibly even from third party).

Reply:
Oh, right, that makes sense :)